### ISSUE TITLE
Example using the OTG female port for USB keyboard?

### ISSUE THREAD

**Issue by bertrandom** [Xinyuan-LilyGO/T-Dongle-ESP32S2#1]
Hi, is there an example of how to hook up a USB keyboard to the USB OTG female port using Arduino?

**Comment by LilyGO** [comment]
You can see the example in the example by installing the ESP32's library.

**Comment by bertrandom** [comment]
![sketch_sep12a | Arduino 1 8 19 2022-09-12 23-27-41](https://user-images.githubusercontent.com/57770/189826347-77133cfe-70e3-416a-a5ee-f4a8ff470e5a.png)
All of the USB examples, including the Keyboard ones, are for programming the male USB connector to show up as a HID device on the computer that it's plugged in to, but none of them demonstrate using the USB female OTG port.

**Comment by sergio-arroyo** [comment]
I join the question.  I've been looking for information about it for a few days.

**Comment by bertrandom** [comment]
Do you have an Arduino example (that's a esp-idf example)? Or an example of using a keyboard HID?

**Comment by lyusupov** [comment]
https://github.com/chegewara/EspTinyUSB

**Comment by harbaum** [comment]
The female port on this board uses the same physical pins as the male connector. This means that you cannot use the host port (the female one) at the same time as the male one connected e.g. to the PC.

**Comment by lyusupov** [comment]
One can use DIP switches to isolate D+/D- signals of the 'male' connector from the ESP32-S2, so both USB ports can be used at the same time:
* male - to supply power
* female - to provide USB-Host function

This is an example of the USB-Host use case:
https://github.com/lyusupov/SoftRF/wiki/WebTop-USB


![](https://github.com/lyusupov/SoftRF/raw/master/documents/images/WebTop-8.jpg)

**Comment by harbaum** [comment]
That's correct. But that means you can only use the male port for power supply then. Data lines are disconnected and you cannot use it as a COM port to the PC.

But the GPIO43 and GPIO44 on the groove are actually UART0 TxD and RxD and it should be possible to use that with a USB TTL serial converter to upload a new sketch and have a serial console.

**Comment by weissm** [comment]
Great stuff. Thanks. I downloaded the WebTop-USB SW and it works nicely. 
To reproduce your results: Any chance to show the related USB code to read external devices?

**Comment by lyusupov** [comment]
@weissm look into ['examples' area of ESP-IDF](https://github.com/Xinyuan-LilyGO/T-Dongle-ESP32S2/issues/1#issuecomment-1257728300) and/or [WebTop](https://github.com/lyusupov/SoftRF/blob/master/software/firmware/source/SkyWatch/Platform_ESP32.cpp)

**Comment by GerOtten** [comment]
I am looking for a device were I can connect a Barcode reader, apply some business rules to the read barcode and send this result to a second usb port that acts like a keyboard interface? Any ideas? So the read barcode is transferred to the right format expected by the software.
